vortex-file/src/read/layouts/flat.rs

```diff
@@ -96,29 +101,30 @@ impl LayoutReader for FlatLayoutReader {
         Ok(())
     }
 
-    fn read_selection(&self, selection: &RowMask) -> VortexResult<Option<BatchRead>> {
+    fn poll_read(&self, selection: &RowMask) -> VortexResult<PollRead<Option<ArrayData>>> {
         if let Some(buf) = self.message_cache.get(&[]) {
             let array = self.array_from_bytes(buf)?;
-            selection
-                .filter_array(array)?
-                .map(|s| {
-                    Ok(BatchRead::Batch(
-                        self.scan
-                            .expr
-                            .as_ref()
-                            .map(|e| e.evaluate(&s))
-                            .transpose()?
-                            .unwrap_or(s),
-                    ))
-                })
-                .transpose()
+
+            if let Some(selected) = selection.filter_array(array)? {
+                let filtered = self
+                    .scan
+                    .expr
+                    .as_ref()
+                    .map(|e| e.evaluate(&selected))
+                    .transpose()?
+                    .unwrap_or(selected);
+                Ok(PollRead::Ready(Some(filtered)))
+            } else {
+                Ok(PollRead::Ready(None))
+            }
```

Comment:
I kinda want to ban transpose; it's so hard to grok.

Reply:
I think all of these long operator chains are too hard to grok tbh. like once you have 6 things being chained, with closures etc. I can't just look at it and immediately figure out what's going on

Reply:
_scala has entered the chat_